Add spanwise shape variables

joanibal opened the pull request:
## Purpose
These new design variables control all the FDD nodes along a line in a chosen spanwise direction

This means that for 2D shape optimization with ADflow, you can use spanwise local design variables to parametrize the 2D shape without the need for added linear constraints. 

I think this is a lot easier to explain with an examples 

Consider the following rectangular FFD with it i,j, and k axes as shown in the figure 
![image](https://user-images.githubusercontent.com/20216619/111381908-b8d58600-867c-11eb-918c-76e953360d5c.png)

If a airfoil surface mesh is embedded in the FFD as shown 
![image](https://user-images.githubusercontent.com/20216619/111382335-3d280900-867d-11eb-8a6e-c1e1bdd3feb1.png)

Then adding spanwise shape variables like this 
```python 
DVGeo.addGeoDVSpanwiseLocal("shape", 'k', lower=-0.5, upper=0.5, axis="z", scale=1.0)

```
would add the 6 design variables that correspond to the six colors in this figure. Each design variables would control the z location of two FFD points. 

![image](https://user-images.githubusercontent.com/20216619/111382696-adcf2580-867d-11eb-9fa9-5668aa5a0366.png)

The test case uses the existing cylinder FFD
![image](https://user-images.githubusercontent.com/20216619/111379859-00a6de00-867a-11eb-9401-2384461f556f.png)


## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
Explain the steps needed to test the new code to verify that it does indeed address the issue and produce the expected behavior.

## Checklist
_Put an `x` in the boxes that apply._

- [no] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
^formatting introduced a large amount of change not relevant to the feature
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation
^ I will after feedback

joanibal commented:
I following the same process that Nick used for `sectionLocal`, but we may want to consider refactoring `DVGeometry` at some point.

anilyil commented:
I don't understand DVGeometry's inner working as well as @marcomangano so I will pass this down to him. The figures in the PR does a great job at explaining what it does, is there a way we can include them in the docs as well?

marcomangano commented:
@joanibal I will carefully review this tomorrow! Just a few points for now:
1. The tests are failing, is that because you trained your `.ref` files locally?
2. Is this for 2D optimization only, or do you envision a broader use?
2.b I need to check the code/examples, but can you select any of `x,y,z` and/or "single rows" of control points? Just to get a grasp of the flexibility of this implementation.

I see the point of having this extra feature but I am trying to figure out how usable it can be for general cases.
Also agree that the figures are a great added value and we should add them to the docs somehow :)

joanibal commented:
@marcomangano 

1) guilty as charged. I did train them locally. Why is that a problem again?

2) I envision broader use. It think it is use for 3D wings too. If you start from a rectangular wing then you can change the shape while keeping a constant airfoil along the span. But I think the best use is for 2D cases.

marcomangano commented:
1. I am not sure if this is an issue for pure-python packages too, but when I added a test (and a locally trained `.ref` file) for ADflow the tests on the docker configuration were not matching the prescribed tolerances on the docker builds. I guess our local machines always have some different settings and configurations that can alter the results, even in the I/O phase. @nwu63 can you explain this more in detail? I have a pretty qualitative overview of the issue. Anyway, not sure if it applies here, will check the tests soon.

2. I see your point, this feature does not actually add new capabilities but rather simplify and make our usual parametrization much leaner. Just the benefit for 2D cases is already enough imho

3. I see, makes sense, will check the details in my review.

Just one last thought before I dive into the code: if not in place already, I would add an `assertAllClose` check that compares FFD locations and sensitivities of the new DV vs the usual multi-DV approach. This way we can catch any discrepancy in future developments.

marcomangano commented on pygeo/DVGeometry.py:
```diff
@@ -2944,6 +3175,87 @@ def _attachedPtJacobian(self, config):
+                N = self.FFD.embededVolumes['child%d_coef'%(iChild)].N
+                self.children[iChild].dCcdXdvl = numpy.zeros((N*3, self.nDV_T))
+
+            iDVSpanwiseLocal = self.nDVSW_count
```
I see most of this comes from `_sectionlocalDVJacobian` but it would be great to have some more inline comments, I am struggling a bit to follow the flow.

joanibal replied:
see new added comments

nwu63 commented:
> I am not sure if this is an issue for pure-python packages too, but when I added a test (and a locally trained `.ref` file) for ADflow the tests on the docker configuration were not matching the prescribed tolerances on the docker builds. I guess our local machines always have some different settings and configurations that can alter the results, even in the I/O phase. @nwu63 can you explain this more in detail? I have a pretty qualitative overview of the issue. Anyway, not sure if it applies here, will check the tests soon.

I am not sure where this discrepancy comes from. Sometimes I've had no problems training stuff locally, other times things behave in really strange ways. In this case though, it appears that the ref file was not added to the PR branch?

marcomangano commented on pygeo/DVGeometry.py:
```diff
@@ -3701,6 +4045,112 @@ def _convertTo1D(value, dim1):
+                self.dv_to_coefs.append(loc_dv_to_coefs)
+
+        print('========')
+        print(self.dv_to_coefs)
```
Is this a debugging printout?

joanibal replied:
yep, It's gone now

marcomangano commented on tests/reg_tests/test_Cylinder.py:
```diff
@@ -142,3 +142,60 @@ def scale_circle(val, geo):
+            leList = [[0, 0, 0 ], [-radius/2, 0, height]]
+            xAxis = [-1, 0, 0]
+            yAxis = [0, 1, 0]
+            DVCon.addLERadiusConstraints(leList, nSpan=5, axis=yAxis,
```
I see this comes from the previous test, but I am not familiar with this script. Why are we using it?

marcomangano commented on tests/reg_tests/test_Cylinder.py:
```diff
@@ -142,3 +142,60 @@ def scale_circle(val, geo):
+                                         chordDir=xAxis, scaled=False)
+            # DVCon.writeTecplot('cylinder_constraints.dat')
+
+            funcs = {}
```
It looks we are only testing the DVCon sens here... which I guess it is pretty much the same than testing the DV sens directly? See my comment before the review about an additional test -  if needed

marcomangano reviewed (commented):
A lot of new code - but luckily pretty close to the previous `GeoDVSectionLocal` implementation. As this is an entirely new feature, I don't see risks of breaking anything (unless the user mixes vars in a weird way?) so I think this is pretty much ready - see the comments on testing though.

I added a bunch of comments, mostly for me to try and better understand how `pyGeo` works. Hope to turn this PR into a chance to improve the code current documentation, any concise comment or explanation is welcome.

Also, I completely agree that at some point `pyGeo` will need some major refactoring

marcomangano commented:
> I am not sure where this discrepancy comes from. Sometimes I've had no problems training stuff locally, other times things behave in really strange ways. In this case though, it appears that the ref file was not added to the PR branch?

L M A O, good catch

bbrelje reviewed (changes requested):
I apologize for the delay in review. I am going to focus on the testing because I don't totally understand the implementation code and I trust you did it generally correctly.

Is there any way of verifying with the cylinder test as written that the feature is working as intended, and not just that it's producing the answer you expect. I am not sure a LERadiusConstraint alone is enough for that. a 2d grid of thickness constraints might be (would want to verify that the numbers are the same in the spanwise direction

There also ought to be a verification that this DV isn't used on a non-2D problem; or the test coverage needs to be extended to cover 3D cases.

joanibal commented on examples/childffd_warning/warning_childFFD.py:
```diff
@@ -0,0 +1,166 @@
+
```
Yes it is. How can I fix this issue. The file isn't shown in my repo or in the changed files here.

marcomangano replied:
@nwu63 suggestions? Maybe if we squash everything this is going to be fine, but the `git blame` might turn a bit inconsistent?

nwu63 replied:
I think this is fine because we'll just squash everything, I wouldn't worry too much about it.

joanibal commented on tests/reg_tests/warning_childFFD.py:
```diff
@@ -1,4 +1,3 @@
-
```
Yes it is. How can I fix this issue. The file isn't shown in my repo or in the changed files here.

joanibal commented on pygeo/DVConstraints.py:
```diff
@@ -3579,6 +3579,27 @@ def _finalize(self):
                 self.ncon += len(cons)
```
Sorry, I'm not following. Did I miss something in this file?

marcomangano replied:
Lol don't worry, it was related to the first spurious commits, I should have deleted this sry

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -144,9 +145,11 @@ def __init__(self, fileName, complex=False, child=False, faceFreeze=None, *args,
         self.nDVG_T = None
         self.nDVL_T = None
         self.nDVSL_T = None
+        self.nDVSW_T = None
```
See the new comments in code to address this

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -690,6 +693,98 @@ def addGeoDVLocal(self, dvName, lower=None, upper=None, scale=1.0,
 
+    def addGeoDVSpanwiseLocal(self, dvName, spanIndex, axis='y', lower=None, upper=None,
+                             scale=1.0, pointSelect=None, volList=None, config=None):
+
```
Added

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -690,6 +693,98 @@ def addGeoDVLocal(self, dvName, lower=None, upper=None, scale=1.0,
+                             scale=1.0, pointSelect=None, volList=None, config=None):
+
+        if type(config) == str:
+                    config = [config]
```
From the doc stings it looks like you can use it to control which design variables are applied to a particular config. For example you could have a landing config where you add trailing edge morphing.

config : str or list
            Define what configurations this design variable will be applied to
            Use a string for a single configuration or a list for multiple
            configurations. The default value of None implies that the design
            variable appies to *ALL* configurations.

marcomangano replied:
I see, I was just not really following the branches of the code where `config` is not the default `None`, I will take a better look. Are we testing this feature? :) 
Anyway, what you say makes sense, but now I am wondering how this is used in a run - i.e. it looks like you can use config to define a DV for a specific `setName` in our multipoint optimization, instead of relying on different DV definitions for different procs as we did for morphing and multi element airfoils in the past. This could make things much easier to set up.

Anyway, this digression is not relevant for this PR so we can talk offline, don't worry about this.

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -690,6 +693,98 @@ def addGeoDVLocal(self, dvName, lower=None, upper=None, scale=1.0,
+            if pointSelect.type != 'ijkBounds':
+                pts, ind = pointSelect.getPoints(self.FFD.coef)
+            else:
+                pts, ind = pointSelect.getPoints_ijk(self)
```
I copied that section from addGeoDVLocal, Should I add that bit of code to both?

marcomangano replied:
Mmmm.. it looks like [`pointSelect`](https://github.com/mdolab/pygeo/blob/50d109ab84a3b1f59cdb53586cbf9d8af895365e/pygeo/geo_utils.py#L1269-L1447) is used to pick a subset of the FFD control points to be used by this variable, and then the `volList` is updated (using the missing line) to include the volumes to which these points belong to.

I am really not sure which of our cases / runs actually end up using this bit of code (I think we usually provide an explicit list of point indexes rather than an object like this) but to me that line looks like it belongs both here and in `addGeoDVLocal`. 
Not sure if this is a small bug or something irrelevant though, and ofc we don't have any tests to check this.

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -690,6 +693,98 @@ def addGeoDVLocal(self, dvName, lower=None, upper=None, scale=1.0,
+        volDVMap = []
+        for ivol in volList:
+
+            spanIdx = ijk_2_idx[spanIndex[ivol]]
```
That is correct. I followed the pattern used for secIndex in `addGeoDVSectionLocal`.

The default behavior is to use the same index for each if one is given
```
         if type(spanIndex) is str:
            spanIndex = [spanIndex]*len(volList)
```

joanibal replied:
I'm not sure if we have an example, I'm unaware of any cases where this is used

marcomangano replied:
yep, a lot of options are untested, but anyway this looks good!

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -1301,6 +1409,10 @@ def update(self, ptSetName, childDelta=True, config=None):
             numpy.put(self.FFD.coef[:, 1], self.ptAttachInd, temp[:, 1])
             numpy.put(self.FFD.coef[:, 2], self.ptAttachInd, temp[:, 2])
 
+        # Now add in the section local DVs
```
done

No particular reason. I saw that Nick had put section local before local, so I though I'd follow suit and append it to the front. I not sure it matters since all the local variables use += when setting the coef of the FFD in their respective __call__ methods.

joanibal commented on pygeo/DVGeometry.py:
```diff
@@ -1955,7 +2122,7 @@ def computeTotalJacobianCS(self, ptSetName, config=None):
 
     def addVariablesPyOpt(self, optProb, globalVars=True, localVars=True,
-                          sectionlocalVars=True, ignoreVars=None, freezeVars=None):
+                          sectionlocalVars=True, spanwiselocalVars=True, ignoreVars=None, freezeVars=None):
```
done

joanibal commented:
I added another test using a 2d grid of thickness constraints as Ben suggested. 
I also addressed the comments Marco had in the code.

It is ready for review again.

nwu63 commented:
Not sure why but there are some merge conflicts, could you resolve those so the newly added test gets run on Azure?

marcomangano reviewed (commented):
Looks good, thanks for the comments! Let's wait for the tests to run

marcomangano commented on pygeo/DVConstraints.py:
```diff
@@ -237,8 +237,8 @@ def setSurface(self, surf, name='default', addToDVGeo=False, DVGeoName='default'
-        if name in self.surfaces.keys():
-            raise KeyError('Surface names must be unique. Repeated surface name: ' + str(name))
+        # if name in self.surfaces.keys():
+        #     raise KeyError('Surface names must be unique. Repeated surface name: ' + str(name))
```
Do we want to remove these lines? We don't check this anymore?

joanibal replied:
oh oops, That's a hack that was supposed to stay on my fork.

marcomangano reviewed (commented):
I think we are good to go! (see comment below). @bbrelje we need your approval of the changes before we merge this.

bbrelje commented on pygeo/DVConstraints.py:
```diff
@@ -254,6 +254,7 @@ def setSurface(self, surf, name='default', addToDVGeo=False, DVGeoName='default'
             elif isinstance(surf, pyGeo): # Assume it's a pyGeo surface
                 p0, v1, v2 = self._generateDiscreteSurface(surf)
             else:
+                import ipdb; ipdb.set_trace()
```
don't think this needs to be here - probably a holdover local change

bbrelje reviewed (changes requested):
I still think it would be best to test the situation where someone tries to use this functionality for a 3D problem (in my view it should throw an error and the test should assert that it throws an error). One minor cleanup item (remove the extra debug statement from DVConstraints).

joanibal commented:
@bbrelje maybe I'm not getting what you mean by 3D problem. In the newest test, these design variables are used for a 3D problem. I agree the main application is 2D, but it isn't wrong to use these design variables for 3D.

nwu63 commented:
> I still think it would be best to test the situation where someone tries to use this functionality for a 3D problem (in my view it should throw an error and the test should assert that it throws an error). One minor cleanup item (remove the extra debug statement from DVConstraints).

My understanding here is that these DVs can be used in 3D cases, and there are tests/docs covering usage. I'm going to merge this but if you think this PR can be improved, please make an issue @bbrelje.
